File: src/stack/src/stack_py/stack_py/mission_manager.py
from enum import IntEnum


class MissionMode(IntEnum):
    IDLE = 0
    WP_FLIGHT = 1
    RESCUE = 2
    INVERSE_WP_FLIGHT = 3
    DROP = 4
    LANDING = 5
    FINISHED = 6
    ABORT = 100


class NodeState(IntEnum):
    IDLE = 0b00
    BUSY = 0b01
    SUCCESS = 0b10
    ABORT = 0b11


class NodeName(IntEnum):
    MISSION = 0
    FLIGHT = 1
    TARGET = 2
    GRIPPER = 3
    VISION = 4
    MARKER = 5
    YOLO = 6
    LOGGER = 7


class MissionManager:
    BITS_PER_NODESTATE = 2
    BITS_PER_NODE = 3

    MASK_NODESTATE = 0b11
    MASK_NODE = 0b111


    def __init__(self):
        self.data = 0


    def set_raw(self, data: int):
        self.data = int(data)


    def raw(self):
        return self.data


    def clear(self):
        self.data = 0


    def set(self, node: NodeName, state: NodeState):

        shift = (int(node) * self.BITS_PER_NODESTATE)
        self.data &= ~(self.MASK_NODESTATE << shift)
        self.data |= (int(state) << shift)


    def get(self, node: NodeName):
        shift = (int(node) * self.BITS_PER_NODESTATE)
        value = (self.data >> shift) & self.MASK_NODESTATE

        return NodeState(value)


    @staticmethod
    def pack(node: NodeName, state: NodeState):
        node_bits = (int(node) & MissionManager.MASK_NODE)
        state_bits = (int(state) & MissionManager.MASK_NODESTATE)

        return (
            (node_bits << MissionManager.BITS_PER_NODESTATE)
            | state_bits
        )


    @staticmethod
    def get_node(cmd: int):
        value = (
            (cmd >> MissionManager.BITS_PER_NODESTATE)
            & MissionManager.MASK_NODE
        )

        return NodeName(value)


    @staticmethod
    def get_command(cmd: int):
        value = (
            cmd
            & MissionManager.MASK_NODESTATE
        )

        return NodeState(value)


    def is_idle(self, node: NodeName):
        return (
            self.get(node)
            == NodeState.IDLE
        )

    def is_busy(self, node: NodeName):
        return (
            self.get(node)
            == NodeState.BUSY
        )

    def is_completed(self, node: NodeName):
        return (
            self.get(node)
            == NodeState.COMPLETED
        )

    def is_error(self, node: NodeName):
        return (
            self.get(node)
            == NodeState.ERROR
        )

File: stack/src/stack_cpp/src/mission_manager.cpp
#include "stack_cpp/mission_manager.h"
#include <iostream>
#include <string>

MissionManager::MissionManager()
  : data(0)
  {}

//sets node state
void MissionManager::set(NodeName node, NodeState state) {
  uint32_t shift =  static_cast<uint32_t>(node) * BITS_PER_NODESTATE;

  data &= ~(MASK_NODESTATE << shift);
  data |= ((static_cast<uint32_t>(state) & MASK_NODESTATE) << shift);
}

void MissionManager::set_mode(MissionMode mode) {
  data &= ~(MASK_MODE << SHIFT_SUMMARY_MODE);
  data |= (static_cast<uint32_t>(mode) & MASK_MODE) << SHIFT_SUMMARY_MODE;
}

MissionMode MissionManager::get_mode() const {
  return static_cast<MissionMode>((data >> SHIFT_SUMMARY_MODE) & MASK_MODE);
}

//gets node state
NodeState MissionManager::get(NodeName node) const {
  uint32_t shift = static_cast<uint32_t>(node)*BITS_PER_NODESTATE;
  
  return static_cast<NodeState>((data >> shift) & MASK_NODESTATE);
}

void MissionManager::clear() {
  data = 0;
}

uint32_t MissionManager::raw() const {
  return data;
}


//pack command
uint32_t MissionManager::pack(NodeName node, NodeState state, MissionMode mode) const {
  uint32_t packet = pack(node, state);
  
  packet |= (static_cast<uint32_t>(mode) & MASK_MODE) << SHIFT_MODE;

  return packet;
}


//pack status 
uint32_t MissionManager::pack(NodeName node, NodeState state) const {
  uint32_t packet = 0;

  packet |= (static_cast<uint32_t>(state) & MASK_NODESTATE) << SHIFT_NODESTATE;
  packet |= (static_cast<uint32_t>(node) & MASK_NODE) << SHIFT_NODE;

  return packet;
}


//gets target node from command
NodeName MissionManager::get_node(uint32_t cmd){
  return static_cast<NodeName>((cmd >> SHIFT_NODE) & MASK_NODE);
}

//gets command from command
NodeState MissionManager::get_command(uint32_t cmd){
  return static_cast<NodeState>((cmd >> SHIFT_NODESTATE) & MASK_NODESTATE);
}

//gets mission mode from command
MissionMode MissionManager::get_mode(uint32_t cmd){
  return static_cast<MissionMode>((cmd >> SHIFT_MODE) & MASK_MODE);
}


std::string MissionManager::get_node_str(NodeName node) const {
  std::string str;
  switch(node){
    case NodeName::MISSION:
      str = "MISSION";
      break;
    case NodeName::FLIGHT:
      str = "FLIGHT";
      break;
    case NodeName::TARGET:
      str = "TARGET";
      break;
    case NodeName::GRIPPER:
      str = "GRIPPER";
      break;
    case NodeName::VISION:
      str = "VISION";
      break;
    case NodeName::MARKER:
      str = "MARKER";
      break;
    case NodeName::YOLO:
      str = "YOLO";
      break;
    case NodeName::LOGGER:
      str = "LOGGER";
      break;
    default:
      str = "[ERROR] Node doesn't exist";
      break;
  }
  return str;
}


std::string MissionManager::get_state_str(NodeState state) const {
  std::string str;
  switch(state){
    case NodeState::IDLE:
      str = "IDLE";
      break;
    case NodeState::BUSY:
      str = "BUSY";
      break;
    case NodeState::SUCCESS:
      str = "SUCCESS";
      break;
    case NodeState::ABORT:
      str = "ABORT";
      break;
    default:
      str = "[ERROR] Node doesn't exist";
      break;
  }
  return str;
}
//TODO
/*
bool isError(int node);
bool isFinished(int node);
void resetNode(int node);
int maxNodes() const;
*/
